### PR TITLE
transpile: Refactor binary operand handling and handle complex numbers

### PR DESCRIPTION
**Pull request by Rua**
Both a refactor and a functional improvement. Complex numbers aren't supported yet, so I haven't included any tests for that, but this is needed to support them later.

### CODE REVIEW

**Review by ahomescu — CHANGES_REQUESTED** [review]
Could you add a test or two for this (unit and/or snapshot)?

**Comment by Rua** [comment]
Not really sure what to test for, that isn't already tested?

**Comment by ahomescu** [comment]
Okay we can land this as is since it's a pretty small change, but it would be nice in the future if changes were split between 1) refactorings (no tests needed), and 2) actual functional changes, that would need a test or two to handle those specific corner cases.

File: c2rust-transpile/src/c_ast/mod.rs [hunk 1]
```diff
@@ -1361,32 +1361,48 @@ impl TypedAstContext {
...
+                                        lhs_type_id
+                                    } else {
+                                        rhs_type_id
+                                    };
+
+                                // For complex arithmetic, complex and real values can be mixed.
+                                // See if a `Complex` version of the new result type exists.
+                                if matches!(result_type_kind, CTypeKind::Complex(..)) {
+                                    let new_result_type_kind =
+                                        CTypeKind::Complex(result_type_id.ctype);
```
Comment (ahomescu) [inline c2rust-transpile/src/c_ast/mod.rs:1564]: What if `result_type_id` is already a `Complex`? Would something like `double f(double _Complex a, double _Complex b) { ... a + b ... }` panic here? That could be a new test.

File: c2rust-transpile/src/c_ast/conversion.rs
Comment (ahomescu) [inline c2rust-transpile/src/c_ast/conversion.rs:452]: All the pullback types are integers, do any of them make sense for complex?

**Review by ahomescu — CHANGES_REQUESTED** [review]

File: c2rust-transpile/src/c_ast/conversion.rs
Comment (Rua) [reply c2rust-transpile/src/c_ast/conversion.rs:452]: They'd probably not be used much, but since C supports them, they probably ought to be here too.

File: c2rust-transpile/src/c_ast/mod.rs (hunk 1 above)
Comment (Rua) [reply c2rust-transpile/src/c_ast/mod.rs:1564]: If it's already a `Complex` then it would either stay the same, or be replaced with the complex version of one of the `PULLBACK_KINDS`, same as happens with other numeric types right now. The code here only applies to binary operators, so it won't modify a function call.

Comment (ahomescu) [reply c2rust-transpile/src/c_ast/mod.rs:1564]: My point was: is it incorrect to create a type like `CTypeKind::Complex(Complex(...))`?

Comment (Rua) [reply c2rust-transpile/src/c_ast/mod.rs:1564]: In C, yes.

Comment (ahomescu) [reply c2rust-transpile/src/c_ast/mod.rs:1564]: Can you check if the following C functions cause transpiler panics?
```C
double _Complex f(double _Complex a, double _Complex b) {
    return a * b;
}

double _Complex g(double a, double _Complex b) {
    return a * b;
}

typedef double real;

double _Complex h(double _Complex a, real b) {
    return a * b;
}
```

Comment (Rua) [reply c2rust-transpile/src/c_ast/mod.rs:1564]: Hmm, it does, and I see the problem now. I managed to fix the first two cases, but the third one still panics because of the `typedef`. No `real _Complex` type exists. I need to think of a way around that...

Comment (Rua) [reply c2rust-transpile/src/c_ast/mod.rs]: Ok, I've completely rewritten that part of the code. It should be good now?
